### DISCUSSION TITLE
记忆化搜索

### DISCUSSION THREAD

**Discussion by 8arry** [CSGrandeur/s-1problem1day1ac#252] [Algorithm]
# 记忆化搜索

### 1.什么是记忆化？

例题优化的斐波那契数列问题

```c++
#include<bits/stdc++.h>
using namespace std;
long long f[100];
long long fib(int n)
{
	if(n==1||n==2)
	return f[n]=1;
	if(f[n])
	return f[n];
	f[n]=fib(n-1)+fib(n-2);
	return f[n];
}
int main()
{
	int n;
	while(~scanf("%d",&n))
	{
	printf("%lld\n",fib(n));
	}
	return 0;
} 
```

[![img](https://camo.githubusercontent.com/a696bc0a6639a3451ba304b8180ca0d5bff438b311b51bce70ea4a2badea5ff8/68747470733a2f2f692e626d702e6f76682f696d67732f323032312f30382f373133643663636165303839646136622e706e67)](https://camo.githubusercontent.com/a696bc0a6639a3451ba304b8180ca0d5bff438b311b51bce70ea4a2badea5ff8/68747470733a2f2f692e626d702e6f76682f696d67732f323032312f30382f373133643663636165303839646136622e706e67)将已经计算过的值储存起来，减少计算次数以达到优化的效果

#### 共同点

1.都有状态转移方程

#### 不同点

1.记忆化搜索自顶向下，动态规划自底向上

#### 模板

```c++
int g[maxn];
int dfs(传入数值)
{
	if(g[规模]!=无效数值)
	return g[规模];
	if(终止条件)
	return 最小子问题的解;
	g[规模] = dfs(缩小规模); 
	return g[规模];
}
int main()
{
	memset(g, 无效数值, sizeof(g));
	dfs();
}
```

#### 方法一

1. 把这道题的 dp 状态和方程写出来
2. 根据它们写出 dfs 函数
3. 添加记忆化数组

#### 方法二

1. 写出这道题的暴搜程序（最好是 [dfs](https://oi-wiki.org/search/dfs/))
2. 将这个 dfs 改成“无需外部变量”的 dfs
3. 添加记忆化数组

#### 01背包问题

•现在 SZTUer 拿出了 x*x* 个迷你装药物（嗑药打人可耻…），准备开始与对手决斗。

•由于迷你装药物每个只能用一次，所以 SZTUer 要谨慎的使用这些药。悲剧的是，用药量没达到最少打败该人所需的属性药药量，则打这个人必输。例如他用 22 个药去打别人，别人却表明 33 个药才能打过，那么相当于你输了并且这两个属性药浪费了。

•现在有 n个对手，给定失败时可获得的经验lose[i]、胜利时可获得的经验win[i]，打败他至少需要的药量use[i]。

•要求求出最大经验(今天五倍经验)

```c++
#include<bits/stdc++.h>  
using namespace std;  
int n,x;  
long long los[1005],win[1005];  
long long s[1005][1005];  
long long med[1005];  
long long dfs(int i,int xx)  
{  
	if(s[i][xx])
	return s[i][xx];
	if(i==n+1)
	return 0;
	if(xx>=med[i])
	s[i][xx]=max(dfs(i+1,xx-med[i])+win[i],dfs(i+1,xx)+los[i]);
	else
	s[i][xx]=dfs(i+1,xx)+los[i];
	return s[i][xx];
}

int main()
{
	memset(s,0,sizeof(s));
	scanf("%d%d",&n,&x);
	for(int i=1;i<=n;i++)
	{
		scanf("%d%d%d",&los[i],&win[i],&med[i]);
	}
	printf("%lld\n",dfs(1,x)*5);
	return 0;
}
```

#### 字符串问题

•输入两串字符串a和b，要求最少的字符操作次数，将字符串a转换为字符串b

•字符操作共有三种

•1.删除一个字符；

•2.插入一个字符；

•3.将一个字符改为另一个字符；

```c++
#include<bits/stdc++.h>
using namespace std;
char a[2005],b[2005];
char a1[2005],b1[2005];
int s[2005][2005];
int dfs(int n,int m)
{
	if(s[n][m])
	return s[n][m];
	if(n==0)
	return s[n][m]=m;
	if(m==0)
	return s[n][m]=n;
	if(a1[n]==b1[m])
	return s[n][m]=dfs(n-1,m-1);
	s[n][m]=min(min(dfs(n-1,m)+1,dfs(n,m-1)+1),dfs(n-1,m-1)+1);
	return s[n][m];
}
int main()
{	
	int n;
	int m;
	memset(s,0,sizeof(s));
	scanf("%s",a);
	scanf("%s",b);
	n=strlen(a);
	m=strlen(b);
	for(int i=1;i<=n;i++)
		a1[i]=a[i-1];
	for(int i=1;i<=m;i++)
		b1[i]=b[i-1];
	printf("%d",dfs(n,m));
	return 0;
}
```

#### 最值问题

•Michael 喜欢滑雪。这并不奇怪，因为滑雪的确很刺激。可是为了获得速度，滑的区域必须向下倾斜，而且当你滑到坡底，你不得不再次走上坡或者等待升降机来载你。Michael 想知道在一个区域中最长的滑坡。区域由一个二维数组给出。数组的每个数字代表点的高度。下面是一个例子：

```c++
#include<bits/stdc++.h>
using namespace std;
int a[202][202],s[202][202];
int dx[4]={1,0,-1,0};
int dy[4]={0,1,0,-1};
int n,m,ans=0;
int dfs(int x,int y)
{
	if(s[x][y]) return s[x][y]; 
	s[x][y]=1;
	for(int i=0;i<4;i++)
	{
		int xx=dx[i]+x;
		int yy=dy[i]+y;
		if(xx>0&&yy>0&&xx<=n&&yy<=m&&a[x][y]>a[xx][yy])
		{
			dfs(xx,yy);
			s[x][y]=max(s[x][y],s[xx][yy]+1);
		}
	}
	return s[x][y];
}
int main()
{
	memset(s,0,sizeof(s));
	scanf("%d%d",&n,&m);
	for(int i=1;i<=n;i++)
	for(int j=1;j<=m;j++)
	scanf("%d",&a[i][j]);
	for(int i=1;i<=n;i++)
	for(int j=1;j<=m;j++)
	ans=max(ans,dfs(i,j));
	printf("%d",ans);
	return 0;
}
```

### 5.题单

| 序号 | 题号         | 标题                   | 题型       | 题目难度 |
| ---- | ------------ | ---------------------- | ---------- | -------- |
| 1    | Poj1579      | Function Run Fun       | 模板题     | 二星     |
| 2    | P2626        | 优化的斐波那契数列问题 | 模板题     | 二星     |
| 3    | IOI1994      | 数字三角形             | 最值问题   | 三星     |
| 4.   | P1802        | 5倍经验日              | 背包问题   | 三星     |
| 5    | P1434        | 滑雪                   | 最值问题   | 三星     |
| 6.   | p2758        | 编辑距离               | 字符串问题 | 四星     |
| 7.   | 蓝桥杯2016省 | 回文字串               | 字符串问题 | 四星     |
| 8.   | HDU 1078     | FatMouse and Cheese    | 最值问题   | 三星     |
| 9.   | HDU 1501     | Zipper                 | 字符串问题 | 四星     |
| 10.  | HDU 1508     | Alphacode              | 字符串问题 | 三星     |
|      |              |                        |            |          |
|      |              |                        |            |          |
|      |              |                        |            |          |
|      |              |                        |            |          |
|      |              |                        |            |          |
|      |              |                        |            |          |
|      |              |                        |            |          |
|      |              |                        |            |          |

**Comment by CSGrandeur** [comment]
典型写法：

```cpp
const int maxn = 1010;
const int maxm = 510;
int dp[maxn][maxm];
void Init()
{
    memset(dp, -1, sizeof(dp));
}

int DPS(int xn, int xm)
{
    if(xn > n || xm > m) {return /* ... */; }    // 数据规模边界
    int &ans = dp[xn][xm];                     // 引用，修改ans 即修改 dp[xn][xm]
    if(ans != -1) return ans;
    if(/* 终点条件 */) {return ans = /*... */ ;}               // 处理终点条件，且同时标记 ans 为终点值，即不再是-1
    // 接下来处理更新逻辑，调用子问题
    ans = 0;   // 重要，设置一个安全的初值，使其为访问过不再是-1
    // 根据题目要求执行子问题，这里是随便举例
    ans = max(ans, DPS(xn + 1, xm + 1));
    ans = max(ans, DPS(xn + 1, xm));
    ans = max(ans, DPS(xn, xm + 1));
    return ans;
}

```
